File: examples/python/decode_morse.py
from __future__ import annotations

import argparse
import json
import math
import wave
from pathlib import Path

import numpy as np
import onnxruntime as ort

MIN_AUDIO_SECONDS = 5.0
MAX_AUDIO_SECONDS = 20.0


def load_metadata(path: Path) -> dict:
    with path.open("r", encoding="utf-8") as file:
        return json.load(file)


def read_wav_mono(path: Path) -> tuple[int, np.ndarray]:
    """Read PCM WAV audio and return mono float32 samples in the range [-1, 1]."""
    with wave.open(str(path), "rb") as wav:
        channels = wav.getnchannels()
        sample_width = wav.getsampwidth()
        sample_rate = wav.getframerate()
        frame_count = wav.getnframes()
        raw = wav.readframes(frame_count)

    if sample_width == 1:
        audio = (np.frombuffer(raw, dtype=np.uint8).astype(np.float32) - 128.0) / 128.0
    elif sample_width == 2:
        audio = np.frombuffer(raw, dtype="<i2").astype(np.float32) / 32768.0
    elif sample_width == 4:
        audio = np.frombuffer(raw, dtype="<i4").astype(np.float32) / 2147483648.0
    else:
        raise ValueError(f"Unsupported PCM sample width: {sample_width} bytes")

    if channels > 1:
        audio = audio.reshape(-1, channels).mean(axis=1)

    return sample_rate, audio.astype(np.float32, copy=False)


def resample_linear(audio: np.ndarray, source_rate: int, target_rate: int) -> np.ndarray:
    """Resample with linear interpolation to keep the example dependency-light."""
    if source_rate == target_rate:
        return audio.astype(np.float32, copy=False)
    if len(audio) == 0:
        return audio.astype(np.float32, copy=False)

    target_length = int(round(len(audio) * target_rate / source_rate))
    source_positions = np.arange(target_length, dtype=np.float64) * source_rate / target_rate
    left = np.floor(source_positions).astype(np.int64)
    right = np.minimum(left + 1, len(audio) - 1)
    fraction = (source_positions - left).astype(np.float32)
    resampled = audio[left] * (1.0 - fraction) + audio[right] * fraction
    return resampled.astype(np.float32, copy=False)


def frequency_bin_range(sample_rate: int, fft_length: int, min_hz: float, max_hz: float) -> tuple[int, int]:
    bin_hz = sample_rate / fft_length
    start_bin = int(math.ceil(min_hz / bin_hz))
    stop_bin = int(math.floor(max_hz / bin_hz)) + 1
    return start_bin, stop_bin


def audio_to_spectrogram(audio: np.ndarray, metadata: dict) -> np.ndarray:
    """Create a [1, 1, time, frequency] float32 spectrogram for the ONNX model."""
    fft_length = int(metadata["fft_length"])
    hop_length = int(metadata["hop_length"])
    sample_rate = int(metadata["sample_rate"])
    min_hz = float(metadata["spectrogram_min_freq_hz"])
    max_hz = float(metadata["spectrogram_max_freq_hz"])
    expected_bins = int(metadata["spectrogram_frequency_bins"])

    if len(audio) < fft_length:
        raise ValueError(f"Audio is too short for fft_length={fft_length}")

    start_bin, stop_bin = frequency_bin_range(sample_rate, fft_length, min_hz, max_hz)
    if stop_bin - start_bin != expected_bins:
        raise ValueError(f"Metadata expects {expected_bins} bins, but the computed range has {stop_bin - start_bin}")

    pad = fft_length // 2
    audio = np.pad(audio, (pad, pad), mode="reflect")
    window = np.hanning(fft_length + 1)[:-1].astype(np.float32)
    frames = 1 + (len(audio) - fft_length) // hop_length
    spectrogram = np.empty((frames, expected_bins), dtype=np.float32)

    for frame_index in range(frames):
        start = frame_index * hop_length
        frame = audio[start : start + fft_length] * window
        spectrum = np.fft.rfft(frame, n=fft_length)
        spectrogram[frame_index] = np.abs(spectrum[start_bin:stop_bin]).astype(np.float32)

    if metadata.get("normalization") == "log1p":
        spectrogram = np.log1p(spectrogram, dtype=np.float32)
    else:
        raise ValueError(f"Unsupported normalization: {metadata.get('normalization')}")

    return spectrogram[np.newaxis, np.newaxis, :, :].astype(np.float32, copy=False)


def greedy_ctc_decode(log_probs: np.ndarray, chars: list[str], blank_index: int) -> str:
    """Collapse a CTC best path by removing blanks and repeated labels."""
    best_path = log_probs[0].argmax(axis=-1)
    decoded: list[str] = []
    previous: int | None = None

    for index in best_path:
        index = int(index)
        if index == blank_index:
            previous = None
            continue
        if index != previous:
            decoded.append(chars[index])
        previous = index

    return "".join(decoded)


def decode(model_path: Path, metadata_path: Path, wav_path: Path) -> str:
    metadata = load_metadata(metadata_path)
    target_rate = int(metadata["sample_rate"])
    source_rate, audio = read_wav_mono(wav_path)
    audio = resample_linear(audio, source_rate, target_rate)

    duration_seconds = len(audio) / target_rate
    if duration_seconds < MIN_AUDIO_SECONDS or duration_seconds > MAX_AUDIO_SECONDS:
        raise ValueError(
            f"Audio duration must be between {MIN_AUDIO_SECONDS:.0f} and {MAX_AUDIO_SECONDS:.0f} seconds "
            f"after resampling to {target_rate} Hz. Got {duration_seconds:.2f} seconds."
        )

    spectrogram = audio_to_spectrogram(audio, metadata)
    session = ort.InferenceSession(str(model_path), providers=["CPUExecutionProvider"])
    outputs = session.run([metadata["onnx_output_name"]], {metadata["onnx_input_name"]: spectrogram})
    return greedy_ctc_decode(outputs[0], list(metadata["chars"]), int(metadata["blank_index"]))


def main() -> None:
    parser = argparse.ArgumentParser(description="Decode Morse code audio with the DeepCW ONNX model.")
    parser.add_argument("--model", type=Path, default=Path("../../model.onnx"), help="Path to model.onnx.")
    parser.add_argument("--metadata", type=Path, default=Path("../../model.onnx.json"), help="Path to model metadata JSON.")
    parser.add_argument("--wav", type=Path, required=True, help="Path to a 5-20 second PCM WAV file.")
    args = parser.parse_args()

    text = decode(args.model, args.metadata, args.wav)
    print(text)


if __name__ == "__main__":
    main()

File: examples/nodejs/decode_morse.mjs
import fs from "node:fs";
import path from "node:path";
import * as ort from "onnxruntime-node";

const MIN_AUDIO_SECONDS = 5.0;
const MAX_AUDIO_SECONDS = 20.0;

function parseArgs() {
  const args = {
    model: "../../model.onnx",
    metadata: "../../model.onnx.json",
    wav: null,
  };

  for (let i = 2; i < process.argv.length; i += 1) {
    const key = process.argv[i];
    const value = process.argv[i + 1];
    if (key === "--model" && value) args.model = value;
    if (key === "--metadata" && value) args.metadata = value;
    if (key === "--wav" && value) args.wav = value;
    if (key.startsWith("--")) i += 1;
  }

  if (!args.wav) {
    throw new Error("Missing required argument: --wav <path-to-5-to-20-second-pcm-wav>");
  }

  return args;
}

function readString(buffer, offset, length) {
  return buffer.toString("ascii", offset, offset + length);
}

function readWavMono(filePath) {
  const buffer = fs.readFileSync(filePath);
  if (readString(buffer, 0, 4) !== "RIFF" || readString(buffer, 8, 4) !== "WAVE") {
    throw new Error("Expected a RIFF/WAVE file.");
  }

  let offset = 12;
  let format = null;
  let dataOffset = -1;
  let dataSize = 0;

  while (offset + 8 <= buffer.length) {
    const chunkId = readString(buffer, offset, 4);
    const chunkSize = buffer.readUInt32LE(offset + 4);
    const chunkData = offset + 8;

    if (chunkId === "fmt ") {
      format = {
        audioFormat: buffer.readUInt16LE(chunkData),
        channels: buffer.readUInt16LE(chunkData + 2),
        sampleRate: buffer.readUInt32LE(chunkData + 4),
        bitsPerSample: buffer.readUInt16LE(chunkData + 14),
      };
    } else if (chunkId === "data") {
      dataOffset = chunkData;
      dataSize = chunkSize;
    }

    offset = chunkData + chunkSize + (chunkSize % 2);
  }

  if (!format || dataOffset < 0) {
    throw new Error("The WAV file is missing a fmt or data chunk.");
  }
  if (format.audioFormat !== 1) {
    throw new Error(`Only PCM WAV is supported by this compact example. Got format ${format.audioFormat}.`);
  }
  if (![8, 16, 32].includes(format.bitsPerSample)) {
    throw new Error(`Unsupported PCM bit depth: ${format.bitsPerSample}`);
  }

  const bytesPerSample = format.bitsPerSample / 8;
  const frameCount = Math.floor(dataSize / (bytesPerSample * format.channels));
  const audio = new Float32Array(frameCount);

  for (let frame = 0; frame < frameCount; frame += 1) {
    let sum = 0;
    for (let channel = 0; channel < format.channels; channel += 1) {
      const sampleOffset = dataOffset + (frame * format.channels + channel) * bytesPerSample;
      if (format.bitsPerSample === 8) sum += (buffer.readUInt8(sampleOffset) - 128) / 128;
      if (format.bitsPerSample === 16) sum += buffer.readInt16LE(sampleOffset) / 32768;
      if (format.bitsPerSample === 32) sum += buffer.readInt32LE(sampleOffset) / 2147483648;
    }
    audio[frame] = sum / format.channels;
  }

  return { sampleRate: format.sampleRate, audio };
}

function resampleLinear(audio, sourceRate, targetRate) {
  if (sourceRate === targetRate) return audio;
  const targetLength = Math.round((audio.length * targetRate) / sourceRate);
  const output = new Float32Array(targetLength);

  for (let i = 0; i < targetLength; i += 1) {
    const sourcePosition = (i * sourceRate) / targetRate;
    const left = Math.floor(sourcePosition);
    const right = Math.min(left + 1, audio.length - 1);
    const fraction = sourcePosition - left;
    output[i] = audio[left] * (1 - fraction) + audio[right] * fraction;
  }

  return output;
}

function frequencyBinRange(sampleRate, fftLength, minHz, maxHz) {
  const binHz = sampleRate / fftLength;
  return {
    startBin: Math.ceil(minHz / binHz),
    stopBin: Math.floor(maxHz / binHz) + 1,
  };
}

function selectedDftMagnitudes(frame, startBin, stopBin) {
  const length = frame.length;
  const output = new Float32Array(stopBin - startBin);

  for (let bin = startBin; bin < stopBin; bin += 1) {
    let real = 0;
    let imaginary = 0;
    for (let n = 0; n < length; n += 1) {
      const angle = (-2 * Math.PI * bin * n) / length;
      real += frame[n] * Math.cos(angle);
      imaginary += frame[n] * Math.sin(angle);
    }
    output[bin - startBin] = Math.hypot(real, imaginary);
  }

  return output;
}

function audioToSpectrogram(audio, metadata) {
  const fftLength = metadata.fft_length;
  const hopLength = metadata.hop_length;
  const expectedBins = metadata.spectrogram_frequency_bins;
  if (audio.length < fftLength) throw new Error(`Audio is too short for fft_length=${fftLength}.`);

  const { startBin, stopBin } = frequencyBinRange(
    metadata.sample_rate,
    fftLength,
    metadata.spectrogram_min_freq_hz,
    metadata.spectrogram_max_freq_hz,
  );
  if (stopBin - startBin !== expectedBins) {
    throw new Error(`Metadata expects ${expectedBins} bins, but computed ${stopBin - startBin}.`);
  }

  const pad = Math.floor(fftLength / 2);
  const padded = new Float32Array(audio.length + pad * 2);
  for (let i = 0; i < pad; i += 1) {
    padded[i] = audio[pad - i];
    padded[pad + audio.length + i] = audio[audio.length - 2 - i];
  }
  padded.set(audio, pad);

  const frames = 1 + Math.floor((padded.length - fftLength) / hopLength);
  const tensorData = new Float32Array(frames * expectedBins);
  const window = new Float32Array(fftLength);
  for (let i = 0; i < fftLength; i += 1) {
    window[i] = 0.5 - 0.5 * Math.cos((2 * Math.PI * i) / fftLength);
  }

  const frame = new Float32Array(fftLength);
  for (let frameIndex = 0; frameIndex < frames; frameIndex += 1) {
    const start = frameIndex * hopLength;
    for (let i = 0; i < fftLength; i += 1) {
      frame[i] = padded[start + i] * window[i];
    }
    const magnitudes = selectedDftMagnitudes(frame, startBin, stopBin);
    for (let bin = 0; bin < expectedBins; bin += 1) {
      tensorData[frameIndex * expectedBins + bin] = Math.log1p(magnitudes[bin]);
    }
  }

  return new ort.Tensor("float32", tensorData, [1, 1, frames, expectedBins]);
}

function greedyCtcDecode(logProbs, chars, blankIndex) {
  const [batch, frames, classes] = logProbs.dims;
  if (batch !== 1) throw new Error(`Expected batch size 1, got ${batch}.`);

  let previous = null;
  let decoded = "";
  for (let frame = 0; frame < frames; frame += 1) {
    let bestIndex = 0;
    let bestValue = -Infinity;
    for (let klass = 0; klass < classes; klass += 1) {
      const value = logProbs.data[frame * classes + klass];
      if (value > bestValue) {
        bestValue = value;
        bestIndex = klass;
      }
    }

    if (bestIndex === blankIndex) {
      previous = null;
    } else {
      if (bestIndex !== previous) decoded += chars[bestIndex];
      previous = bestIndex;
    }
  }

  return decoded;
}

async function main() {
  const args = parseArgs();
  const modelPath = path.resolve(process.cwd(), args.model);
  const metadataPath = path.resolve(process.cwd(), args.metadata);
  const wavPath = path.resolve(process.cwd(), args.wav);
  const metadata = JSON.parse(fs.readFileSync(metadataPath, "utf8"));
  const { sampleRate, audio } = readWavMono(wavPath);
  const resampled = resampleLinear(audio, sampleRate, metadata.sample_rate);
  const durationSeconds = resampled.length / metadata.sample_rate;

  if (durationSeconds < MIN_AUDIO_SECONDS || durationSeconds > MAX_AUDIO_SECONDS) {
    throw new Error(
      `Audio duration must be between ${MIN_AUDIO_SECONDS} and ${MAX_AUDIO_SECONDS} seconds after resampling. ` +
        `Got ${durationSeconds.toFixed(2)} seconds.`,
    );
  }

  const spectrogram = audioToSpectrogram(resampled, metadata);
  const session = await ort.InferenceSession.create(modelPath);
  const outputs = await session.run({ [metadata.onnx_input_name]: spectrogram });
  const decoded = greedyCtcDecode(outputs[metadata.onnx_output_name], metadata.chars, metadata.blank_index);
  console.log(decoded);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
